workflow/controller/operator.go

```diff
@@ -806,6 +809,10 @@ func (woc *wfOperationCtx) persistUpdates(ctx context.Context) {
 			woc.log.WithError(err).Warn("failed to delete task-results")
 		}
 	}
+	// If FinalizerArtifactGC exists, requeue to make sure artifact GC can execute.
+	if woc.wf.Status.Fulfilled() && slices.Contains(wf.GetFinalizers(), common.FinalizerArtifactGC) {
+		woc.requeue()
```

Comment:
would this result in an infinite loop where the workflow always in the wfqueue?

Reply:
FinalizerArtifactGC should be removed after gc completed.

Reply:
maybe we should theoretically requeue if there's any Finalizer, not just Artifact GC?

Reply:
I agree

Reply:
also, how do you feel about adding to this to the block above `if woc.wf.Status.Fulfilled() {` as a nested if statement?

Reply:
If for all Finalizers instead of just FinalizerArtifactGC, if woc.wf.Status.Fulfilled can be removed

Reply:
That seems reasonable and nice and clean. Only thing now is I want to make sure we're not getting ourselves into a situation of constant queuing and cpu usage that we wouldn't have had before. Can you see how this will affect that? Will we be continually processing the same workflow over and over? I wonder if we should instead use the `requeueAfter()` function?

Reply:
Based on the [log](https://github.com/argoproj/argo-workflows/issues/12836#issuecomment-2016683367) and my understanding of how artifactGC works, I think the issue is something else.

Currently, the finalizer does have a chance to be removed, see pseudo code below
```
// operate()
if wf.status.fulfilled
   garbageCollectArtifacts
      create WorkflowArtifactGCTask
      create artifact gc pod
          loop through WorkflowArtifactGCTask
          patch WorkflowArtifactGCTask for each deletion
// controller()
WorkflowArtifactGCTaskInformer.
   on WorkflowArtifactGCTask Update
      wfqueue.AddRateLimited(key) // this requeue wf to remove finalizer
```

Reply:
Looking at the test failure linked above, is it possible that the only issue is that right after [deleting](https://github.com/argoproj/argo-workflows/blob/main/test/e2e/artifacts_test.go#L207) the Workflow, the Artifact is [checked](https://github.com/argoproj/argo-workflows/blob/main/test/e2e/artifacts_test.go#L213) for being deleted, so maybe the Workflow Controller didn't have a chance to complete PodGC yet?

Reply:
On version 3.5.5, I've noticed that when I stop & delete workflows in the UI before they are complete, finalizers aren't removed and the workflow gets stuck (but artifact gc succeeds). I have to do a manual wf edit to clear them. This feels like the same issue.

@juliev0 Looking at some of the test failures linked, I'm noticing the following:
```
Waiting 1m30s for workflows {{ } workflows.argoproj.io/test metadata.name=artgc-dag-wf-stopped-pod-gc-on-pod-completion-qrlp5 false false   <nil> 0 }
    when.go:356: client rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline
```
Indicating a rate limit issue with calling `wfList, err := w.client.List(ctx, listOptions)`. If we're getting a rate limit error, and execution moves on to the artifact presence check too soon, then yes, it could be that the controller didn't have a chance to finish PodGC yet.

@shuangkun @juliev0 per my first comment here, I still think there's an issue that needs to be solved (and I think your proposed solution could be sufficient). As @tczhao mentioned, requeueing could be an issue, but only if finalizers are never removed for some reason.

If adding rate limiting for `WaitForWorkflowList` resolves these transient test issues, then we still need to modify the `TestStoppedWorkflow` test to ensure that finalizers are removed.

Reply:
(Sorry, I just realized my last comment said "PodGC" - I meant to say "ArtifactGC")

> If we're getting a rate limit error, and execution moves on to the artifact presence check too soon, then yes, it could be that the controller didn't have a chance to finish PodGC yet. <-- I assume you also mean "ArtifactGC"

@Garett-MacGowan Thanks for tying that together. Actually, in my original statement, I didn't realize that [this  WaitForWorkflowDeletion()](https://github.com/argoproj/argo-workflows/blob/main/test/e2e/artifacts_test.go#L208) call essentially waits for the finalizer to have been removed, so the logic does make sense, except for the rate limiting case I guess.

Reply:
> On version 3.5.5, I've noticed that when I stop & delete workflows in the UI before they are complete, finalizers aren't removed and the workflow gets stuck (but artifact gc succeeds).

That's interesting that it sounds like there's also a real issue. It's notable what @tczhao pointed out that the WorkflowArtifactGCTaskInformer should [requeue](https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/controller.go#L1500) when the ArtifactGCTask has changed, which should cause the Workflow to be processed and the WorkflowArtifactGCTasks listed and read [here](https://github.com/argoproj/argo-workflows/blob/main/workflow/controller/artifact_gc.go#L623) and then the Finalizer removed. Maybe there's some race condition in there that could prevent that? 

I guess at the end of the day we need to determine if there's any harm in this change. If we change `woc.requeue()` to `woc.requeueAfter(delay)` and reduce the immediacy of the requeue, then maybe it's generally a good thing that we regularly revisit Workflows that still have Finalizers just in case?

Reply:
Yes, true. I guess there is something broken as far as finalizer logic if we get into that scenario.